testing/describes.js

```diff
@@ -0,0 +1,220 @@
+ * @param {{
+ *   fakeClock: (boolean|undefined),
+ * }} spec
+ * @param {function({
+ *   sandbox: !Sandbox,
+ *   clock: (!FakeClock|undefined),
+ * })} fn
+ */
+export function sandboxed(name, spec, fn) {
+  return describe(name, function() {
+
+    const test = Object.create(null);
+
+    beforeEach(() => {
+      test.sandbox = sinon.sandbox.create();
+      if (spec.fakeClock) {
+        test.clock = test.sandbox.useFakeTimers();
+      }
+    });
+
+    afterEach(() => {
+      test.sandbox.restore();
+      delete test.sandbox;
+      if (test.clock) {
+        delete test.clock;
+      }
+    });
+
+    describe(SUB, function() {
+      fn.call(this, test);
+    });
+  });
+}
+
+
+/**
+ * A test with a fake window.
+ * @param {string} name
+ * @param {{win: !FakeWindowSpec}} spec
+ * @param {function({
+ *   win: !FakeWindow,
+ * })} fn
+ */
+export function fakeWin(name, spec, fn) {
+  return describe(name, function() {
+
+    const test = Object.create(null);
+
+    beforeEach(() => {
+      test.win = new FakeWindow(spec);
+    });
+
+    afterEach(() => {
+      // TODO(dvoytenko): test that window is returned in a good condition.
+      delete test.win;
+    });
+
+    describe(SUB, function() {
+      fn.call(this, test);
+    });
+  });
+}
+
+
+/**
+ * A test with a fake window.
+ * @param {string} name
+ * @param {{
+ *   fakeRegisterElement: (boolean|undefined),
+ *   amp: (!AmpTestSpec|undefined),
+ * }} spec
+ * @param {function({
+ *   sandbox: !Sandbox,
+ *   clock: (!FakeClock|undefined),
+ * })} fn
+ */
+export function realWin(name, spec, fn) {
+  return describe(name, function() {
+
+    const test = Object.create(null);
+
+    beforeEach(() => {
+      return new Promise(function(resolve, reject) {
+        const iframe = document.createElement('iframe');
+        iframe.name = 'test_' + iframeCount++;
+        iframe.srcdoc = '<!doctype><html><head>' +
+            '<style>.-amp-element {display: block;}</style>' +
+            '<body style="margin:0"><div id=parent></div>';
+        iframe.onload = function() {
+          const win = iframe.contentWindow;
+          test.iframe = iframe;
+          test.win = win;
+
+          // Flag as being a test window.
+          win.AMP_TEST_IFRAME = true;
+
+          doNotLoadExternalResourcesInTest(win);
+
+          if (spec.fakeRegisterElement) {
+            win.customElements = new FakeCustomElements(win);
+          }
+
+          if (spec.amp) {
+            ampSetup(test, win, spec.amp);
+          }
+          resolve();
+        };
+        iframe.onerror = reject;
+        document.body.appendChild(iframe);
+      });
+    });
+
+    afterEach(() => {
+      ampDestroy(test);
+      // TODO(dvoytenko): test that window is returned in a good condition.
+      if (test.iframe.parentNode) {
+        test.iframe.parentNode.removeChild(test.iframe);
+      }
+      delete test.iframe;
+      delete test.win;
+    });
+
+    describe(SUB, function() {
```

Comment:
No, but I have to call `describe` here to get tests-own beforeEach and afterEach called and if I call `describe`, it must have a name, even if it's just a space.